File: app/src/main/java/com/solt9029/editmasterandroid/model/Resource.java
package com.solt9029.editmasterandroid.model;

public class Resource<T> {
    public T data = null;
    public boolean isLoading = false;
    public Throwable error = null;

    public Resource() {

    }

    private Resource(T data, boolean isLoading, Throwable error) {
        this.data = data;
        this.isLoading = isLoading;
        this.error = error;
    }

    public static <T> Resource<T> startLoading(T data) {
        return new Resource<>(data, true, null);
    }

    public static <T> Resource<T> finishLoadingSuccess(T data) {
        return new Resource<>(data, false, null);
    }

    public static <T> Resource<T> finishLoadingFailure(Throwable error) {
        return new Resource<>(null, false, error);
    }
}

File: app/src/main/java/com/solt9029/editmasterandroid/viewmodel/Resource.kt
package com.solt9029.editmasterandroid.viewmodel

class Resource<T>() {
    var data: T? = null
    var isLoading = false
    var error: Throwable? = null

    private constructor(data: T?, isLoading: Boolean, error: Throwable?) : this() {
        this.data = data
        this.isLoading = isLoading
        this.error = error
    }

    companion object {

        fun <T> startLoading(data: T?): Resource<T> {
            return Resource(data, true, null)
        }

        fun <T> finishLoadingSuccess(data: T?): Resource<T> {
            return Resource(data, false, null)
        }

        fun <T> finishLoadingFailure(error: Throwable?): Resource<T> {
            return Resource(null, false, error)
        }
    }
}
